cmd/rehearse/main.go

```diff
@@ -0,0 +1,222 @@
+package main
+
+import (
+	"flag"
+	"fmt"
+	"os"
+	"strings"
+
+	// TODO: Solve this properly
+	// "github.com/davecgh/go-spew/spew"
+	"github.com/getlantern/deepcopy"
+	// "github.com/ghodss/yaml"
+	"github.com/sirupsen/logrus"
+
+	pjapi "k8s.io/test-infra/prow/apis/prowjobs/v1"
+	pjclientset "k8s.io/test-infra/prow/client/clientset/versioned"
+	"k8s.io/test-infra/prow/config"
+	prowflagutil "k8s.io/test-infra/prow/flagutil"
+	"k8s.io/test-infra/prow/github"
+	"k8s.io/test-infra/prow/pjutil"
+
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+)
+
+func getJobsToExecute(ca *config.Agent) config.JobConfig {
+	var youHaveOneJob config.Presubmit
+	for _, job := range ca.Config().Presubmits["openshift/ci-operator"] {
+		if job.Name == "pull-ci-openshift-ci-operator-master-build" {
+			youHaveOneJob = job
+			break
+		}
+	}
+
+	jobs := config.JobConfig{
+		Presubmits: map[string][]config.Presubmit{"openshift/ci-operator": []config.Presubmit{youHaveOneJob}},
+	}
+	return jobs
+}
+
+func makeRehearsalPresubmit(source *config.Presubmit, repo string, prNumber int) (*config.Presubmit, error) {
+	var rehearsal config.Presubmit
+	deepcopy.Copy(&rehearsal, source)
+
+	rehearsal.Name = fmt.Sprintf("rehearse-%d-%s", prNumber, source.Name)
+	rehearsal.Context = fmt.Sprintf("ci/rehearse/%s/%s", repo, strings.TrimPrefix(source.Context, "ci/prow/"))
+
+	if len(source.Spec.Containers) != 1 {
+		return nil, fmt.Errorf("Cannot rehearse jobs with more than 1 container in Spec")
+	}
+	container := source.Spec.Containers[0]
+
+	if len(container.Command) != 1 || container.Command[0] != "ci-operator" {
+		return nil, fmt.Errorf("Cannot rehearse jobs that have Command different from simple 'ci-operator'")
+	}
+
+	for _, arg := range container.Args {
+		if strings.HasPrefix(arg, "--git-ref") || strings.HasPrefix(arg, "-git-ref") {
+			return nil, fmt.Errorf("Cannot rehearse jobs that call ci-operator with '--git-ref' arg")
+		}
+	}
+
+	if len(source.Branches) != 1 {
+		return nil, fmt.Errorf("Cannot rehearse jobs that run over multiple branches")
```

Comment:
Why?

Reply:
Because it does not fit into my current program structure where one input `JobConfig` makes one rehearsal `JobConfig`. It seem's we do not have mutlibranch jobs at all, because we force by-branch sharding of job config files anyway, so meh. But if desired, this won't be hard to fix, provided we can actually extract some meaningful branch name from the regex.

Also. any ideas about what to do with the theoretical jobs withhout the `Brancher`, meaning they run on all branches? We could always ask GitHub for branches for that repo, but I wonder if it's worth the effort to handle all these cases we do not use in practice.

Reply:
For those we should just run `master` branch. Seems fine for us to assume one branch

Reply:
Still, it does not hurt to check. I also realized this will be run over candidate jobs, where people can put all kinds of "invalid" jobs, so being defensive really does not hurt - if some one makes a PR with a multibranch job, we will tell him no to do it, but it's better that the rehearsal job simply refuses to rehearse it instead of blowing up on it.

Reply:
If someone will do that, the `determine-jobs` job will fail, right? In any case, shall we make the `pj-rehearser` tool to wait for the the `determine-jobs` to be completed and then continue? or even better we can run those checks in `pj-rehearser` binary instead and get rid of the `determinize-prow-jobs` binary.